theories/newth/MapAux.ec

```diff
@@ -0,0 +1,146 @@
+(*---------------------- Auxiliary Lemmas on Maps ----------------------*)
+
+prover [""].
+
+require import AllCore SmtMap FSet StdOrder.
+import IntOrder.
+
+lemma get_none (m : ('a, 'b) fmap, x : 'a) :
+  x \notin m => m.[x] = None.
+proof. by rewrite domE. qed.
+
+lemma get_some (m : ('a, 'b) fmap, x : 'a) :
+  x \in m => m.[x] = Some (oget m.[x]).
+proof. move=> /domE; by case m.[x]. qed.
+
+lemma set_same (m : ('a, 'b) fmap, x : 'a) :
+  x \in m => m.[x <- oget m.[x]] = m.
+proof.
+move=> x_in_m.
+apply fmap_eqP => y.
+case (y = x) => [->> | ne_y_x].
+by rewrite get_set_sameE get_some.
+by rewrite get_setE ne_y_x.
+qed.
+
+lemma set_eq (m : ('a, 'b) fmap, x : 'a, y : 'b) :
+  m.[x] = Some y => m.[x <- y] = m.
+proof.
+move=> m_get_x_eq_y.
+have x_in_m : x \in m by rewrite domE m_get_x_eq_y.
+have -> : y = oget m.[x] by rewrite m_get_x_eq_y oget_some.
+by rewrite set_same.
+qed.
+
+lemma frng_set (m : ('a, 'b) fmap, x : 'a, y : 'b) :
+  frng m.[x <- y] = frng (rem m x) `|` fset1 y.
```

Comment:
State the lemma for `rng` and lift it to `frng`.

Reply:
I don't see the point of doing that for an equational rewriting lemma like this. The point of frng as opposed to rng is to allow such rewriting. Unless I'm missing something, it's messier to state using rng given the right-hand-side involving `|`.